File: frontend/src/api/docqa.js
/**
 * Document Q&A Chat API Client
 */
import apiClient from './client';

function asArray(payload, keys = []) {
  if (Array.isArray(payload)) return payload;
  if (!payload || typeof payload !== 'object') return [];
  for (const key of keys) {
    if (Array.isArray(payload[key])) return payload[key];
  }
  return [];
}

/**
 * Create a Q&A session
 */
export async function createSession(name) {
  const response = await apiClient.post('/docqa/sessions', { name });
  return response.data;
}

/**
 * List all Q&A sessions
 */
export async function listSessions({ limit, offset } = {}) {
  const params = {};
  if (limit != null) params.limit = limit;
  if (offset != null) params.offset = offset;
  const response = await apiClient.get('/docqa/sessions', { params });
  const payload = response.data;
  if (Array.isArray(payload)) {
    return { sessions: payload, total: payload.length };
  }
  if (payload && typeof payload === 'object') {
    const sessions = asArray(payload, ['sessions', 'items', 'results']);
    return { ...payload, sessions, total: payload.total ?? sessions.length };
  }
  return { sessions: [], total: 0 };
}

/**
 * Get a Q&A session
 */
export async function getSession(sessionId) {
  const response = await apiClient.get(`/docqa/sessions/${sessionId}`);
  return response.data;
}

/**
 * Delete a Q&A session
 */
export async function deleteSession(sessionId) {
  const response = await apiClient.delete(`/docqa/sessions/${sessionId}`);
  return response.data;
}

/**
 * Add document to session
 */
export async function addDocument(sessionId, { name, content, pageCount }) {
  const response = await apiClient.post(`/docqa/sessions/${sessionId}/documents`, {
    name,
    content,
    page_count: pageCount,
  });
  return response.data;
}

/**
 * Remove document from session
 */
export async function removeDocument(sessionId, documentId) {
  const response = await apiClient.delete(`/docqa/sessions/${sessionId}/documents/${documentId}`);
  return response.data;
}

/**
 * Ask a question
 */
export async function askQuestion(sessionId, { question, includeCitations = true, maxResponseLength = 2000 }) {
  const response = await apiClient.post(`/docqa/sessions/${sessionId}/ask`, {
    question,
    include_citations: includeCitations,
    max_response_length: maxResponseLength,
  });
  return response.data;
}

/**
 * Get chat history
 */
export async function getChatHistory(sessionId, limit = 50) {
  const response = await apiClient.get(`/docqa/sessions/${sessionId}/history`, {
    params: { limit },
  });
  return response.data;
}

/**
 * Clear chat history
 */
export async function clearHistory(sessionId) {
  const response = await apiClient.delete(`/docqa/sessions/${sessionId}/history`);
  return response.data;
}

/**
 * Submit feedback for a message
 * @param {string} sessionId - Session ID
 * @param {string} messageId - Message ID
 * @param {object} feedback - Feedback data { feedbackType: 'helpful' | 'not_helpful', comment?: string }
 */
export async function submitFeedback(sessionId, messageId, { feedbackType, comment }) {
  const response = await apiClient.post(
    `/docqa/sessions/${sessionId}/messages/${messageId}/feedback`,
    {
      feedback_type: feedbackType,
      comment,
    }
  );
  return response.data;
}

/**
 * Regenerate a response for a message
 * @param {string} sessionId - Session ID
 * @param {string} messageId - Message ID
 * @param {object} options - Options { includeCitations?: boolean, maxResponseLength?: number }
 */
export async function regenerateResponse(sessionId, messageId, { includeCitations = true, maxResponseLength = 2000 } = {}) {
  const response = await apiClient.post(
    `/docqa/sessions/${sessionId}/messages/${messageId}/regenerate`,
    {
      include_citations: includeCitations,
      max_response_length: maxResponseLength,
    }
  );
  return response.data;
}

export default {
  createSession,
  listSessions,
  getSession,
  deleteSession,
  addDocument,
  removeDocument,
  askQuestion,
  getChatHistory,
  clearHistory,
  submitFeedback,
  regenerateResponse,
};

File: backend/app/api/routes/docqa.py
"""API routes for Document Q&A Chat."""
from __future__ import annotations

import logging
import threading
from typing import Optional
from pydantic import BaseModel, Field
from fastapi import APIRouter, Depends, Query, Request, HTTPException

logger = logging.getLogger("neura.api.docqa")

from backend.app.services.security import require_api_key
from backend.app.services.docqa.service import DocumentQAService
from backend.app.schemas.docqa import AskRequest, FeedbackRequest, RegenerateRequest

router = APIRouter(dependencies=[Depends(require_api_key)])


class CreateSessionRequest(BaseModel):
    name: str = Field(..., min_length=1, max_length=200)


class AddDocumentRequest(BaseModel):
    # Limit content to 500KB to prevent memory exhaustion
    # For larger documents, use file upload with chunked processing
    name: str = Field(..., min_length=1, max_length=200)
    content: str = Field(..., min_length=10, max_length=500 * 1024)
    page_count: Optional[int] = None


_lock = threading.Lock()
_docqa_service: DocumentQAService | None = None


def get_service() -> DocumentQAService:
    """Return a singleton DocumentQAService instance."""
    global _docqa_service
    if _docqa_service is None:
        with _lock:
            if _docqa_service is None:
                _docqa_service = DocumentQAService()
    return _docqa_service


@router.post("/sessions")
async def create_session(
    payload: CreateSessionRequest,
    request: Request,
    svc: DocumentQAService = Depends(get_service),
):
    """Create a new Q&A session."""
    correlation_id = getattr(request.state, "correlation_id", None)
    session = svc.create_session(
        name=payload.name,
        correlation_id=correlation_id,
    )
    return {"status": "ok", "session": session.model_dump(mode="json"), "correlation_id": correlation_id}


@router.get("/sessions")
async def list_sessions(
    request: Request,
    limit: int = Query(50, ge=1, le=500),
    offset: int = Query(0, ge=0),
    svc: DocumentQAService = Depends(get_service),
):
    """List all Q&A sessions."""
    correlation_id = getattr(request.state, "correlation_id", None)
    sessions = svc.list_sessions()
    page = sessions[offset:offset + limit]
    return {
        "status": "ok",
        "sessions": [s.model_dump(mode="json") for s in page],
        "total": len(sessions),
        "correlation_id": correlation_id,
    }


@router.get("/sessions/{session_id}")
async def get_session(
    session_id: str,
    request: Request,
    svc: DocumentQAService = Depends(get_service),
):
    """Get a Q&A session by ID."""
    correlation_id = getattr(request.state, "correlation_id", None)
    session = svc.get_session(session_id)

    if not session:
        raise HTTPException(status_code=404, detail="Session not found")

    return {"status": "ok", "session": session.model_dump(mode="json"), "correlation_id": correlation_id}


@router.delete("/sessions/{session_id}")
async def delete_session(
    session_id: str,
    request: Request,
    svc: DocumentQAService = Depends(get_service),
):
    """Delete a Q&A session."""
    correlation_id = getattr(request.state, "correlation_id", None)
    success = svc.delete_session(session_id)

    if not success:
        raise HTTPException(status_code=404, detail="Session not found")

    return {"status": "ok", "deleted": True, "correlation_id": correlation_id}


@router.post("/sessions/{session_id}/documents")
async def add_document(
    session_id: str,
    payload: AddDocumentRequest,
    request: Request,
    svc: DocumentQAService = Depends(get_service),
):
    """Add a document to a Q&A session."""
    correlation_id = getattr(request.state, "correlation_id", None)
    document = svc.add_document(
        session_id=session_id,
        name=payload.name,
        content=payload.content,
        page_count=payload.page_count,
        correlation_id=correlation_id,
    )

    if not document:
        raise HTTPException(status_code=404, detail="Session not found")

    return {"status": "ok", "document": document.model_dump(mode="json"), "correlation_id": correlation_id}


@router.delete("/sessions/{session_id}/documents/{document_id}")
async def remove_document(
    session_id: str,
    document_id: str,
    request: Request,
    svc: DocumentQAService = Depends(get_service),
):
    """Remove a document from a session."""
    correlation_id = getattr(request.state, "correlation_id", None)
    success = svc.remove_document(session_id, document_id)

    if not success:
        raise HTTPException(status_code=404, detail="Session or document not found")

    return {"status": "ok", "removed": True, "correlation_id": correlation_id}


@router.post("/sessions/{session_id}/ask")
async def ask_question(
    session_id: str,
    payload: AskRequest,
    request: Request,
    svc: DocumentQAService = Depends(get_service),
):
    """Ask a question about the documents in a session."""
    correlation_id = getattr(request.state, "correlation_id", None)

    session = svc.get_session(session_id)
    if not session:
        raise HTTPException(status_code=404, detail="Session not found")

    response = svc.ask(session_id, payload, correlation_id)

    if not response:
        raise HTTPException(status_code=500, detail="Failed to process question")

    return {
        "status": "ok",
        "response": response.model_dump(mode="json"),
        "correlation_id": correlation_id,
    }


@router.post("/sessions/{session_id}/messages/{message_id}/feedback")
async def submit_feedback(
    session_id: str,
    message_id: str,
    payload: FeedbackRequest,
    request: Request,
    svc: DocumentQAService = Depends(get_service),
):
    """Submit feedback for a chat message."""
    correlation_id = getattr(request.state, "correlation_id", None)

    session = svc.get_session(session_id)
    if not session:
        raise HTTPException(status_code=404, detail="Session not found")

    message = svc.submit_feedback(
        session_id,
        message_id,
        payload,
        correlation_id,
    )

    if not message:
        raise HTTPException(status_code=404, detail="Message not found")

    return {
        "status": "ok",
        "message": message.model_dump(mode="json"),
        "correlation_id": correlation_id,
    }


@router.post("/sessions/{session_id}/messages/{message_id}/regenerate")
async def regenerate_response(
    session_id: str,
    message_id: str,
    payload: RegenerateRequest,
    request: Request,
    svc: DocumentQAService = Depends(get_service),
):
    """Regenerate a response for a message."""
    correlation_id = getattr(request.state, "correlation_id", None)

    session = svc.get_session(session_id)
    if not session:
        raise HTTPException(status_code=404, detail="Session not found")

    try:
        response = svc.regenerate_response(
            session_id,
            message_id,
            payload,
            correlation_id,
        )
    except RuntimeError as exc:
        raise HTTPException(status_code=500, detail="Document Q&A operation failed") from exc
    except Exception as exc:
        logger.error("Response regeneration failed: %s", exc, exc_info=True)
        raise HTTPException(
            status_code=500,
            detail="Response regeneration failed.",
        ) from exc

    if not response:
        raise HTTPException(status_code=404, detail="Message not found")

    return {
        "status": "ok",
        "response": response.model_dump(mode="json"),
        "correlation_id": correlation_id,
    }


@router.get("/sessions/{session_id}/history")
async def get_chat_history(
    session_id: str,
    request: Request,
    limit: int = Query(50, ge=1, le=500),
    svc: DocumentQAService = Depends(get_service),
):
    """Get chat history for a session."""
    correlation_id = getattr(request.state, "correlation_id", None)

    session = svc.get_session(session_id)
    if not session:
        raise HTTPException(status_code=404, detail="Session not found")

    messages = svc.get_chat_history(session_id, limit)

    return {
        "status": "ok",
        "messages": [m.model_dump(mode="json") for m in messages],
        "count": len(messages),
        "correlation_id": correlation_id,
    }


@router.delete("/sessions/{session_id}/history")
async def clear_chat_history(
    session_id: str,
    request: Request,
    svc: DocumentQAService = Depends(get_service),
):
    """Clear chat history for a session."""
    correlation_id = getattr(request.state, "correlation_id", None)

    success = svc.clear_history(session_id)
    if not success:
        raise HTTPException(status_code=404, detail="Session not found")

    return {"status": "ok", "cleared": True, "correlation_id": correlation_id}
